src/diffusers/pipelines/flux/pipeline_flux.py

```diff
@@ -375,16 +430,73 @@ def encode_prompt(
         dtype = self.text_encoder.dtype if self.text_encoder is not None else self.transformer.dtype
         text_ids = torch.zeros(prompt_embeds.shape[1], 3).to(device=device, dtype=dtype)
 
-        return prompt_embeds, pooled_prompt_embeds, text_ids
+        if do_true_cfg and negative_prompt is not None:
+            return (
+                prompt_embeds,
+                pooled_prompt_embeds,
+                text_ids,
+                negative_prompt_embeds,
+                negative_pooled_prompt_embeds,
+            )
+        else:
+            return prompt_embeds, pooled_prompt_embeds, text_ids, None, None
+
+    def encode_image(self, image, device, num_images_per_prompt):
+        dtype = next(self.image_encoder.parameters()).dtype
+
+        if not isinstance(image, torch.Tensor):
+            image = self.feature_extractor(image, return_tensors="pt").pixel_values
+
+        image = image.to(device=device, dtype=dtype)
+        image_embeds = self.image_encoder(image).image_embeds
+        image_embeds = image_embeds.repeat_interleave(num_images_per_prompt, dim=0)
+        return image_embeds
+
+    def prepare_ip_adapter_image_embeds(
+        self, ip_adapter_image, ip_adapter_image_embeds, device, num_images_per_prompt
+    ):
+        image_embeds = []
+        if ip_adapter_image_embeds is None:
+            if not isinstance(ip_adapter_image, list):
+                ip_adapter_image = [ip_adapter_image]
+
+            if len(ip_adapter_image) != len(self.transformer.encoder_hid_proj.image_projection_layers):
+                raise ValueError(
+                    f"`ip_adapter_image` must have same length as the number of IP Adapters. Got {len(ip_adapter_image)} images and {len(self.transformer.encoder_hid_proj.image_projection_layers)} IP Adapters."
+                )
+
+            for single_ip_adapter_image, image_proj_layer in zip(
+                ip_adapter_image, self.transformer.encoder_hid_proj.image_projection_layers
+            ):
+                single_image_embeds = self.encode_image(single_ip_adapter_image, device, 1)
+
+                image_embeds.append(single_image_embeds[None, :])
+                image_embeds = self.transformer.encoder_hid_proj(image_embeds)
```

Comment:
ohhh! so we are generating ip hidden states directly here, so there is no reason to load ip-adapter into transformer, (for ip-adapter in unet, we only encode the image inside pipeline and let the unet generate the ip-adapter hidden states https://github.com/huggingface/diffusers/blob/f781b8c30c4d70fbf0afcc9799c7f9e9693b2921/src/diffusers/models/unets/unet_2d_condition.py#L1026C19-L1026C35) 

in that case, can we try the design similar to redux? I've been meant to move to that design because it is more "modular". It is also probably a lot easier to implement, we don't need to convert the weights, and it will just be a prior pipeline that encode image and generate the `joint_attention_kwargs` ,joint_https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/flux/pipeline_flux_prior_redux.py

Reply:
Thanks, that design would be easier. SD3.5 IP Adapter is doing it the same as here and has an issue with `enable_sequential_cpu_offload` and `enable_model_cpu_offload` https://github.com/huggingface/diffusers/pull/9987#issuecomment-2552672602 and https://github.com/huggingface/diffusers/pull/9987#issuecomment-2552693475 which I haven't tested for Flux yet, might be specific to Siglip in SD3.5 case but this design will resolve it either way.

Reply:
Though we still need to have `load_ip_adapter` for the weights in the custom attention processor and load some weights from the same safetensors for the separate pipeline.

Reply:
Update: SD3.5 IP Adapter was doing it like this, it's confirmed that was the issue with `enable_model_cpu_offload`
https://github.com/huggingface/diffusers/pull/9987#issuecomment-2553716852 so I've moved usage of `encoder_hid_proj` into `FluxTransformer2DModel`. The issue with `enable_sequential_cpu_offload` is specific to Siglip, Flux IP Adapter uses CLIP Vision so should be ok with `enable_sequential_cpu_offload`.
I'll still look into the design like redux.

Reply:
redux pipeline is a separate pipeline that contains the model components( in this case it will just the image projection layers and image_encoder) https://github.com/huggingface/diffusers/blob/1826a1e7d31df48d345a20028b3ace48f09a4e60/src/diffusers/pipelines/flux/pipeline_flux_prior_redux.py#L124

you won't need load weights, though we still need to set attention processor

Reply:
The attention processor also has weights, `to_k_ip` and `to_v_ip`.

Reply:
ohhh that's true!

Reply:
also we started to make model-specific attentions (mochi has one now) maybe we can make one for ip-adapter so we don't load weights into the attention processor 

still need to load a weights so let's think about the design a bit, but we want something that's very easy to transite into modular system

Reply:
like this https://github.com/huggingface/diffusers/blob/074798b2997a6f1a329924b400a0db924e8e6735/src/diffusers/models/attention_processor.py#L909

sana has their own attention class too now

Reply:
Yes I think passing ip_key and ip_value will work, nice idea!